[arm64e] Sign pointers to ObjC atomic property getter/setter copy helper functions passed to objc_copyCppObjectAtomic

ahatanaka opened the pull request:
This fixes a runtime crash. objc_copyCppObjectAtomic calls the helper functions indirectly through the function pointers and expects the function pointers to be signed using key 0.

ahatanaka committed:
[arm64e] Sign pointers to ObjC atomic property getter/setter copy helper
functions passed to objc_copyCppObjectAtomic

This fixes a runtime crash. objc_copyCppObjectAtomic calls the helper
functions indirectly through the function pointers and expects the
function pointers to be signed using key 0.

ahatanaka commented:
We started seeing the crash after https://github.com/apple/llvm-project/commit/28f7087c9163f66b07e17bd47a1bfb5d9448467. Prior to that commit, clang wasn't emitting calls to `objc_copyCppObjectAtomic`, which was a bug.

`objc_copyCppObjectAtomic` is defined here: https://opensource.apple.com/source/objc4/objc4-818.2/runtime/objc-accessors.mm.auto.html

ahmedbougacha commented:
Setting aside non-trivial-c-struct-property, I see we already emit calls to `objc_copyCppObjectAtomic ` for non-trivial c++ properties, in `ptrauth-property-object-reference.mm`.  However, the passed helper function pointer is already signed there, presumably thanks to `getFunctionPointer` in the non-early-return in `GenerateObjCAtomicGetterCopyHelperFunction` ?  I'm assuming that does generic function-pointer signing;  should it be made aware of the dedicated schema?  I'm not sure the distinction would ever matters in practice (maybe that just means we don't need the additional schema)

rjmccall commented:
> Setting aside non-trivial-c-struct-property, I see we already emit calls to `objc_copyCppObjectAtomic ` for non-trivial c++ properties, in `ptrauth-property-object-reference.mm`. However, the passed helper function pointer is already signed there, presumably thanks to `getFunctionPointer` in the non-early-return in `GenerateObjCAtomicGetterCopyHelperFunction` ? I'm assuming that does generic function-pointer signing; should it be made aware of the dedicated schema? I'm not sure the distinction would ever matters in practice (maybe that just means we don't need the additional schema)

Yeah, I think there are two compiler bugs here:

- We're not signing the callback at all for non-trivial C structs.  We should be able to do that pretty easily without having to add a new constant schema.
- There's a latent bug where we're signing as if the parameter was typed `void (T *, const T *)`, but it's actually `void (void *, const void *)`; current function type discrimination doesn't diversify by pointee types, but we should be doing the right thing.

ahatanaka committed:
Call getFunctionPointer instead of using a new schema for the helper
functions

Pass the correct function type ('void (void *, const void *)') to
calls to getFunctionPointer in both C and C++ mode.

rjmccall reviewed (approved):
Thanks, LGTM
